lib/cluster/membership.c

```diff
@@ -1001,3 +1014,154 @@ crm_terminate_member_no_mainloop(int nodeid, const char *uname, int *connection)
 {
     return stonith_api_kick(nodeid, uname, 120, TRUE);
 }
+
+static void
+known_peer_cache_refresh_helper(xmlNode *xml_node, void *user_data)
+{
+    const char *id = crm_element_value(xml_node, XML_ATTR_ID);
+    const char *uname = crm_element_value(xml_node, XML_ATTR_UNAME);
+    crm_node_t * node =  NULL;
+
+    CRM_CHECK(id != NULL && uname !=NULL, return);
+    node = g_hash_table_lookup(crm_known_peer_cache, id);
+
+    if (node == NULL) {
+        node = calloc(1, sizeof(crm_node_t));
+        if (node == NULL) {
+            errno = -ENOMEM;
+            return;
+        }
+
+        node->uname = strdup(uname);
+        node->uuid = strdup(id);
+        if (node->uname == NULL || node->uuid == NULL) {
+            free(node);
+            errno = -ENOMEM;
+            return;
+        }
+
+        g_hash_table_replace(crm_known_peer_cache, node->uuid, node);
+
+    } else if (is_set(node->flags, crm_node_dirty)) {
+        if (safe_str_neq(uname, node->uname)) {
+            free(node->uname);
+            node->uname = strdup(uname);
+            CRM_ASSERT(node->uname != NULL);
+        }
+
+        /* Node is in cache and hasn't been updated already, so mark it clean */
+        clear_bit(node->flags, crm_node_dirty);
+    }
+
+}
+
+#define XPATH_MEMBER_NODE_CONFIG \
+    "//" XML_TAG_CIB "/" XML_CIB_TAG_CONFIGURATION "/" XML_CIB_TAG_NODES \
+    "/" XML_CIB_TAG_NODE "[not(@type) or @type='member']"
```

Comment:
I didn't dig into xmlXPathEvalExpression() ,  but it appears  xpath "/cib/.." doesn't match anything although it's supposed to ... Not sure if all the xpaths for remote nodes start with "//" because of the same issue.

Reply:
Oh right, the cib is wrapped in some IPC result tags. It's fine to leave it "//". Ideally, we'd be more careful on the calling end to find and pass the cib tag, but it's not important enough to worry about now.